lib/position/bk.js

```diff
@@ -211,36 +211,69 @@ function horizontalCompaction(g, layering, root, align, reverseSep) {
   // coordinates. The second sweep removes unused space by moving blocks to the
   // greatest coordinates without violating separation.
   var xs = {},
-      blockG = buildBlockGraph(g, layering, root, reverseSep);
-
-  // First pass, assign smallest coordinates via DFS
-  var visited = {};
-  function pass1(v) {
+      visited = {},
+      blockG = buildBlockGraph(g, layering, root, reverseSep),
+      borderType = reverseSep ? "borderLeft" : "borderRight";
+
+  // This function handles iterating over the graph without adding functions
+  // calls to the stack. The argument, helper, returns an array of functions.
+  // These returned functions are the recursive and base function calls. By
+  // returning the functions and calling them here, we do not add additional
+  // function calls to the stack. If recursive case generates additional
+  // functions calls, they will be prepended to the array.
+  function iterate(helper, v) {
     if (!_.has(visited, v)) {
       visited[v] = true;
-      xs[v] = _.reduce(blockG.inEdges(v), function(max, e) {
-        pass1(e.v);
-        return Math.max(max, xs[e.v] + blockG.edge(e));
-      }, 0);
+      var traversalFuncs = helper(v);
+
+      while(traversalFuncs.length) {
+        var funcsOrX = traversalFuncs.shift()();
+        if (funcsOrX instanceof Array) {
+          traversalFuncs = funcsOrX.concat(traversalFuncs);
+        }
+      }
     }
   }
-  _.forEach(blockG.nodes(), pass1);
 
-  var borderType = reverseSep ? "borderLeft" : "borderRight";
+  // pass1 maps over the inEdges of a node and creates a recursive function that
+  // will be called by iterate. At the end of those functions, there is another
+  // function that will set xs[v] with the smallest coordinates.
+  function pass1(v) {
+    return _.map(blockG.inEdges(v).reverse(), function(e) {
+      return function() { return pass1(e.v);};
+    }).concat([function () {
+      xs[v] = _.reduce(blockG.inEdges(v), function(acc, e) {
+        return Math.max(acc, xs[e.v] + blockG.edge(e));
+      }, 0);
+      return xs[v];
+    }]);
+  }
```

Comment:
I tried adding whitespace in various places but it didn't seem to make it clearer. Probably just leave it as-is.